eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxowSpec/Valid.hs

```diff
@@ -0,0 +1,323 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+{-# OPTIONS_GHC -Wno-unused-imports #-}
+
+module Test.Cardano.Ledger.Alonzo.Imp.UtxowSpec.Valid (spec) where
+
+import Cardano.Ledger.Alonzo.Core (
+  -- AlonzoEraScript (..),
+  AlonzoEraTx (..),
+  AlonzoEraTxBody (..),
+  Era,
+  EraCrypto,
+  EraGov,
+  EraTx,
+  EraTxBody,
+  EraTxOut,
+  ShelleyEraTxCert,
+  Tx,
+  TxBody,
+  TxOut,
+  addrTxWitsL,
+  bodyTxL,
+  datsTxWitsL,
+  feeTxBodyL,
+  hashScript,
+  inputsTxBodyL,
+  mkBasicTx,
+  mkBasicTxBody,
+  mkBasicTxWits,
+  outputsTxBodyL,
+  rdmrsTxWitsL,
+  scriptTxWitsL,
+  witsTxL,
+ )
+import Cardano.Ledger.Alonzo.TxWits (Redeemers (..))
+import Cardano.Ledger.Coin (Coin (..))
+import Cardano.Ledger.Plutus.Language (SLanguage (..))
+import Cardano.Ledger.Shelley.LedgerState (
+  UTxOState (..),
+  curPParamsEpochStateL,
+  esLStateL,
+  lsUTxOStateL,
+  nesEsL,
+  smartUTxOState,
+  utxosUtxoL,
+ )
+import Data.Default.Class (Default (..))
+import qualified Data.Map.Strict as Map
+import qualified Data.Sequence.Strict as SSeq
+import qualified Data.Set as Set
+import Lens.Micro ((&), (.~))
+import Lens.Micro.Mtl (use)
+import qualified PlutusLedgerApi.V1 as PV1
+import Test.Cardano.Ledger.Alonzo.ImpTest (
+  AlonzoEraImp,
+  ImpTestM,
+  ImpTestState,
+  getUTxO,
+  getsNES,
+  impMakeUTxOState,
+  impNESL,
+  submitTxAnn_,
+ )
+import Test.Cardano.Ledger.Core.KeyPair (mkWitnessVKey, mkWitnessesVKey)
+import Test.Cardano.Ledger.Imp.Common
+
+spec ::
+  forall era.
+  ( AlonzoEraTx era
+  , AlonzoEraTxBody era
+  , AlonzoEraImp era
+  ) =>
+  SpecWith (ImpTestState era)
+spec = do
+  let
+    tx `shouldProduceState` expectedM = do
+      expected <- expectedM
+      submitTxAnn_ "Submit a transaction with an expected ledger state" tx
+      actual <- getsNES $ nesEsL . esLStateL . lsUTxOStateL
+      actual `shouldBe` expected
```

Comment:
This is one of the worst parts of the previous tests. They compare the full state, rather than verify that expected parts of the state have been correctly changed.

We can talk more about it, but in a nutshell tests should check only the relevant part, eg: if we want to test that a transaction spends an output, we apply a transaction and check that the said output was removed from the state.

Reply:
I had already planned to approach this differently, as we discussed. For now, I've updated this module to contain even more minimal placeholders.